File: app/src/main/java/ua/hospes/rtm/ui/race/detail/RaceItemDetailPresenter.java
package ua.hospes.rtm.ui.race.detail;

import javax.inject.Inject;

import ua.hospes.rtm.core.ui.BasePresenter;
import ua.hospes.rtm.utils.RxUtils;

/**
 * @author Andrew Khloponin
 */
public class RaceItemDetailPresenter extends BasePresenter<RaceItemDetailContract.View> {
    private final RaceItemDetailInteractor interactor;

    @Inject
    RaceItemDetailPresenter(RaceItemDetailInteractor interactor) {
        this.interactor = interactor;
    }

    @Override
    public void attachView(RaceItemDetailContract.View view) {
        super.attachView(view);
    }

    @Override
    public void detachView() {
        super.detachView();
        RxUtils.unsubscribe(this);
    }


    void listenRaceItem(int raceItemId) {
        RxUtils.manage(this, interactor.listenRaceItem(raceItemId)
                .compose(RxUtils.applySchedulers())
                .subscribe(raceItem -> {
                    getView().onUpdateRaceItem(raceItem);
                    if (raceItem == null) return;
                    listenSessions(raceItem.getTeam().getId());
                }, Throwable::printStackTrace));
    }

    private void listenSessions(int teamId) {
        RxUtils.manage(this, interactor.listenSessions(teamId)
                .compose(RxUtils.applySchedulers())
                .subscribe(sessions -> getView().onUpdateSessions(sessions), Throwable::printStackTrace));
    }
}

File: app/src/main/java/ua/hospes/rtm/ui/race/detail/RaceItemDetailPresenter.kt
package ua.hospes.rtm.ui.race.detail

import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import ua.hospes.rtm.core.Presenter
import ua.hospes.rtm.domain.race.RaceRepository
import ua.hospes.rtm.domain.sessions.SessionsRepository
import javax.inject.Inject

@Suppress("EXPERIMENTAL_API_USAGE")
internal class RaceItemDetailPresenter @Inject constructor(
        private val raceRepo: RaceRepository,
        private val sessionRepo: SessionsRepository
) : Presenter<RaceItemDetailContract.View>() {

    override fun onError(throwable: Throwable) = view?.onError(throwable) ?: Unit
    override fun onUnexpectedError(throwable: Throwable) = view?.onError(throwable) ?: Unit


    fun setRaceItemId(id: Long) {
        launch { raceRepo.listen(id).collect { view?.onRaceItem(it) } }
        launch { sessionRepo.listenByRaceId(id).collect { view?.onSessions(it) } }
    }
}